File: Easy/0110_balanced-binary-tree/balanced-binary-tree.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func check(node *TreeNode) int {
	if node == nil {
		return 0
	}

	left := check(node.Left)
	if left == -1 {
		return -1
	}

	right := check(node.Right)
	if right == -1 {
		return -1
	}

	if abs(left - right) > 1 {
		return -1
	}
	return max(left, right) + 1
}

func isBalanced(root *TreeNode) bool {
  return check(root) != -1
}

func buildTree(data string) *TreeNode {
	if data == "" {
		return nil
	}

	values := strings.Split(data, ",")
	if len(values) == 0 || values[0] == "null" {
		return nil
	}

	val, err := strconv.Atoi(values[0])
	if err != nil {
		return nil
	}
	root := &TreeNode{Val: val}

	queue := []*TreeNode{root}
	i := 1

	for len(queue) > 0 && i < len(values) {
		node := queue[0]
		queue = queue[1:]

		if i < len(values) && values[i] != "null" {
			if v, err := strconv.Atoi(values[i]); err == nil {
				node.Left = &TreeNode{Val: v}
				queue = append(queue, node.Left)
			}
		}
		i++

		if i < len(values) && values[i] != "null" {
			if v, err := strconv.Atoi(values[i]); err == nil {
				node.Right = &TreeNode{Val: v}
				queue = append(queue, node.Right)
			}
		}
		i++
	}
	return root
}

func main() {
	tests := []string{
		"3,9,20,null,null,15,7",
		"1,2,2,3,3,null,null,4,4",
		"",
		"1,2,2,3,null,null,3,4,null,null,4",
	}

	for _, tc := range tests {
		tree := buildTree(tc)
		fmt.Printf("---\ntree: [%s]\nisBalanced: %t\n", tc, isBalanced(tree))
	}
}

File: Easy/0110_balanced-binary-tree/balanced-binary-tree.py
#!/bin/python3

from typing import List, Optional
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def check(self, node: Optional[TreeNode]) -> int:
        if not node:
            return 0

        left = self.check(node.left)
        if left == -1:
            return -1

        right = self.check(node.right)
        if right == -1:
            return -1

        if abs(left - right) > 1:
            return -1
        return max(left, right) + 1

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self.check(root) != -1

def buildTree(data: str) -> TreeNode:
    if not data:
        return None

    values = data.split(",")
    if not values or values[0] == "null":
        return None
    
    root = TreeNode(int(values[0]))
    queue = deque([root])
    i = 1

    while queue and i < len(values):
        node = queue.popleft()
        if i < len(values) and values[i] != "null":
            node.left = TreeNode(int(values[i]))
            queue.append(node.left)
        i += 1

        if i < len(values) and values[i] != "null":
            node.right = TreeNode(int(values[i]))
            queue.append(node.right)
        i += 1
    return root


def main():
    tests: List[str]=[
        "3,9,20,null,null,15,7",
        "1,2,2,3,3,null,null,4,4",
        "",
        "1,2,2,3,null,null,3,4,null,null,4",
    ]

    for tc in tests:
        tree = buildTree(tc)
        print(f"---\ntree=[{tc}]\nisBalanced: {Solution().isBalanced(tree)}")

if __name__ == "__main__":
    main()
